xcode/Subconscious/Shared/Components/Deck/CardStackView.swift

```diff
@@ -0,0 +1,305 @@
+                    Text(string)
+                }
+                .foregroundStyle(.secondary)
+            }
+            
+        }
+        .blendMode(blendMode)
+        .padding(DeckTheme.cardPadding)
+        .allowsHitTesting(false)
+        .background(color)
+        .cornerRadius(DeckTheme.cornerRadius)
+    }
+}
+
+enum CardDragGestureProgress {
+    case inactive
+    case started
+    case changed
+}
+
+struct CardStack: View {
+    private static let swipeActivationThreshold = 128.0
+    private static let swipeThrowDistance = 1024.0
+    
+    var deck: [CardModel]
+    var current: Int
+    
+    var onSwipeRight: (CardModel) -> Void
+    var onSwipeLeft: (CardModel) -> Void
+    var onSwipeStart: () -> Void
+    var onSwipeAbandoned: () -> Void
+    var onCardTapped: (CardModel) -> Void
+        
+    // Use a dictionary of offsets so that we can animate two cards at once during the transition.
+    // This dictionary is frequently cleared during the gesture lifecycle.
+    @State var offsets: [CardModel:CGSize] = [:]
+    @GestureState private var gestureState: CardDragGestureProgress = .inactive
+    @Environment(\.colorScheme) var colorScheme
+    
+//    func indexOf(card: CardModel) -> Int? {
+//        return deck.firstIndex(where: { $0.id == card.id })
+//    }
+    
+    func offset(`for`: CardModel) -> CGSize {
+        offsets[`for`] ?? CGSize.zero
+    }
+    
+    func offset(for index: Int) -> CGSize {
+        if index < 0 || index >= deck.count {
+            return CGSize.zero
+        }
+        
+        return offset(for: deck[index])
+    }
+    
+    // Ranges from -1 (skip) to 1 (choose)
+    var swipeProgress: CGFloat {
+        return (
+            offset(
+                for: current
+            ).width / Self.swipeActivationThreshold
+        ).clamp(
+            min: -1.0,
+            max: 1.0
+        )
+    }
+    
+    func stackFactor(for index: Int) -> CGFloat {
+        return max(0, CGFloat(index - current)) / 16.0 - abs(swipeProgress) / 64.0
+    }
+    
+    func rotation(for index: Int, stackFactor: CGFloat) -> Double {
+        let fundamental = 6.0*Double(sin(CGFloat(stackFactor - 0.1) * 40))
+        let harmonic = 2.0*Double(sin((stackFactor - 0.45) * 10))
+        
+        let result = (fundamental + harmonic)
+        
+        return result + (stackFactor * swipeProgress * 5)
+    }
+    
+    var body: some View {
+        VStack {
+            Spacer()
+            GeometryReader { geo in
+                ZStack {
+                    let deck = Array(deck.enumerated())
+                    ForEach(deck, id: \.element.id) { index, card in
+                        if (index >= current - 1 && index < current + 3) {
+                            VStack {
+                                Spacer()
+                                
+                                let stackFactor = stackFactor(for: index)
+                                CardView(entry: card)
```

Comment:
I plan to break this out into modular chunks eventually, but while we're still churning on styles I think it's best to leave it as-is.